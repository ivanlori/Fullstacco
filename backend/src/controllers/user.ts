import bcrypt from 'bcrypt'
import { NextFunction, Request, Response } from 'express'

import User from '../models/user'
import { handleErrorStatus } from './utils'

export const getUsers = (req: Request, res: Response, next: NextFunction) => {
	User.find().then((users) => {
		const usersModified = users.map((user) => {
			return {
				id: user._id,
				isActive: user.isActive,
				lastname: user.lastname,
				name: user.name,
				username: user.username,
				email: user.email,
				role: user.role,
				createdAt: user.createdAt,
				updatedAt: user.updatedAt,
			}
		})
		res.status(200).json([
			...usersModified
		])
	}).catch((err) => handleErrorStatus(err, next))
}

export const getUser = (req: Request, res: Response, next: NextFunction) => {
	User.findById(req.params.id).then((user) => {
		res.status(200).json({
			id: user._id,
			isActive: user.isActive,
			lastname: user.lastname,
			name: user.name,
			username: user.username,
			email: user.email,
			role: user.role,
			createdAt: user.createdAt,
			updatedAt: user.updatedAt,
		})
	}).catch((err) => handleErrorStatus(err, next))
}

export const updateUser = (req: Request, res: Response, next: NextFunction) => {
	User.findByIdAndUpdate(req.params.id, req.body).then(() => {
		res.sendStatus(201)
	}).catch((err) => handleErrorStatus(err, next))
}

export const deleteUser = (req: Request, res: Response, next: NextFunction) => {
	User.findByIdAndDelete(req.params.id).then(() => {
		res.sendStatus(201)
	}).catch((err) => handleErrorStatus(err, next))
}

export const createUser = (req: Request, res: Response, next: NextFunction) => {
	const {
		name,
		lastname,
		username,
		email,
		password,
		emailConfirmed,
		isActive,
		role,
	} = req.body

	bcrypt.hash(password, 12).then((hashedPassword) => {
		const user = new User({
			name,
			lastname,
			username,
			email,
			password: hashedPassword,
			emailConfirmed,
			isActive,
			role,
			createdAt: new Date(),
			updatedAt: null
		})

		return user.save()
	}).then((result) => {
		return res.status(201).json({
			userId: result._id
		})
	}).catch((err) => handleErrorStatus(err, next))
}