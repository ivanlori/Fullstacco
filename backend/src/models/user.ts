import mongoose from 'mongoose'

const Schema = mongoose.Schema

export interface IUser {
	_id: string
	name: string,
	lastname: string,
	username: string,
	email: string,
	password: string,
	emailConfirmed: string,
	isActive: boolean,
	role: number,
	createdAt: Date,
	updatedAt: Date,
	photoUrl: string
	resetToken: string | null,
	resetTokenExpiration: Date | undefined,
	save: () => void
}

const userSchema = new Schema({
	name: {
		type: String,
		required: true,
	},
	lastname: {
		type: String,
		required: true
	},
	username: {
		type: String,
		required: true
	},
	email: {
		type: String,
		required: true
	},
	password: {
		type: String,
		required: true
	},
	emailConfirmed: {
		type: Boolean,
		required: true
	},
	isActive: {
		type: Boolean,
		default: true
	},
	role: Number,
	createdAt: {
		type: Date,
		default: Date.now()
	},
	photoUrl: {
		type: String,
		required: false
	},
	updatedAt: Date,
	resetToken: String,
	resetTokenExpiration: Date,
})

export default mongoose.model('User', userSchema)
