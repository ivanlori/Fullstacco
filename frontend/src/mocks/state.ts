import { configureStore } from "@reduxjs/toolkit"

import rootReducer from '../store'

export const createTestStore = () => {
	const store = configureStore({
		reducer: rootReducer
	})
	return store
}