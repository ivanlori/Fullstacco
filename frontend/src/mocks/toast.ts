import { IToastState } from "components/Toast/store/Toast.model"

const mockedToast: IToastState = {
	text: 'some text',
	style: 'success',
}

export default mockedToast
