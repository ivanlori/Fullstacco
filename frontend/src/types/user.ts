import * as types from '../pages/user/store/User.types'

export interface IUserState {
	email: string
	name: string
	lastname: string
	username: string
	password?: string
	updatedAt?: string
	createdAt?: string
	id?: string
	isActive: boolean
	role: number | undefined
	emailConfirmed: boolean
}

export type IRoleSelection = {
	value: number | undefined,
	label: string | undefined
}

export interface emailModel {
	type: typeof types.SET_EMAIL;
	payload: string;
}

export interface idUserModelAction {
	type: typeof types.SET_ID_USER;
	payload: string;
}

export interface userModelAction {
	type: typeof types.SET_USER;
	payload: IUserState;
}

export type ActionType = emailModel | idUserModelAction | userModelAction
