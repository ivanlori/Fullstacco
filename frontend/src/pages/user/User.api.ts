import axios from 'axios'

import { BASE_API } from 'config'
import { getToken, handleError } from 'utils/utils'

import { IUserState } from './store/User.models'

export const createUser = async (data: IUserState) => {
	try {
		return await axios.post(`${BASE_API}/users/create`, data, {
			headers: {
				'Authorization': `Bearer ${getToken()}`
			}
		})
	} catch (err) {
		return handleError(err)
	}
}

export const getUsers = async () => {
	try {
		const { data } = await axios.get(`${BASE_API}/users/`, {
			headers: {
				'Authorization': `Bearer ${getToken()}`
			}
		})

		return data
	} catch (err) {
		return handleError(err)
	}
}

export const updateUser = async (id: string | undefined, data: IUserState) => {
	try {
		return await axios.patch(`${BASE_API}/users/${id}`, data, {
			headers: {
				'Authorization': `Bearer ${getToken()}`
			}
		})
	} catch (err) {
		return handleError(err)
	}
}

export const getUser = async (id: string | null) => {
	try {
		const { data } = await axios.get(`${BASE_API}/users/${id}`, {
			headers: {
				'Authorization': `Bearer ${getToken()}`
			}
		})
		return data
	} catch (err) {
		return handleError(err)
	}
}

export const deleteUser = async (id: string | undefined) => {
	try {
		return await axios.delete(`${BASE_API}/users/${id}`, {
			headers: {
				'Authorization': `Bearer ${getToken()}`
			}
		})
	} catch (err) {
		return handleError(err)
	}
}