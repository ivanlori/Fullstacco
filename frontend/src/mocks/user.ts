import { IUserState } from "types/user"

export const mockedAdminUser: IUserState = {
	email: 'user@email.com',
	name: 'Adam',
	lastname: 'Smith',
	username: 'Adam123',
	password: '123',
	createdAt: 'date',
	updatedAt: 'date',
	id: '123',
	role: 0,
	isActive: false,
	emailConfirmed: false,
}

export const mockedAccountUser: IUserState = {
	...mockedAdminUser,
	role: 1,
}
