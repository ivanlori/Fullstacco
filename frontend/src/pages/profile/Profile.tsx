import { ReactElement } from 'react'

import classNames from 'classnames'
import { useIntl } from 'react-intl'
import { useSelector } from 'react-redux'

import { IState } from 'types/state'

import styles from './Profile.module.css'

const Profile = (): ReactElement => {
	const user = useSelector((state: IState) => state.user)
	const { formatMessage } = useIntl()

	const activeStyle = classNames('rounded-full w-4 h-4 mr-2', {
		"bg-green_primary": user.isActive,
		"bg-magenta_dark": !user.isActive
	})

	return (
		<div className="lg:pl-24">
			<div className="xl:mx-auto flex" >
				<div className="flex flex-col w-full">
					<div className="bg-white text-gray_900 rounded-xl p-5 m-10">
						<div className="flex">
							<div className={styles.Img}></div>
							<div className="ml-4 flex flex-col">
								<span className="flex items-center text-xl">
									<span className={activeStyle}>
										{user.isActive}
									</span>
									<span>{user.username}</span>
								</span>
								<span>
									<span className="text-gray_500">
										{formatMessage({ id: "fullname" })}:
									</span>
									{' '}
									<span>{user.name} {user.lastname}</span>
								</span>
								<span>
									<span className="text-gray_500">
										{formatMessage({ id: "email" })}:
									</span>
									{' '}
									<span>{user.email}</span>
								</span>
								<span>
									<span className="text-gray_500">
										{formatMessage({ id: 'role' })}: {' '}
									</span>
									<span>
										{user.role === 0 ? 'Admin' : 'Account'}
									</span>
								</span>
							</div>
						</div>
					</div>
				</div>
			</div>
		</div>
	)
}

export default Profile
