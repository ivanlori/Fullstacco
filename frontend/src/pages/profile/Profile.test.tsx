import '@testing-library/jest-dom'
import { MemoryRouter } from 'react-router-dom'

import { mockedAdminUser } from 'mocks/user'
import {
	renderWithProviders
} from 'utils/testing-library-utils'

import Profile from './Profile'

describe('<Profile />', () => {

	test('should match snapshot', () => {

		const { container } = renderWithProviders(
			<MemoryRouter>
				<Profile />
			</MemoryRouter>,
			{
				preloadedState: {
					profile: mockedAdminUser
				},
				withIntl: true
			}
		)

		expect(container).toMatchSnapshot()
	})
})
