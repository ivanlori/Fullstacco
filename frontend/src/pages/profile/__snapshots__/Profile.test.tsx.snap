// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Profile /> should match snapshot 1`] = `
<div>
  <div
    class="lg:pl-24"
  >
    <div
      class="xl:mx-auto flex"
    >
      <div
        class="flex flex-col w-full"
      >
        <div
          class="bg-white text-gray_900 rounded-xl p-5 m-10"
        >
          <div
            class="flex"
          >
            <div
              class="w-1/5"
            >
              <label
                for="photo"
              >
                <div
                  class="PhotoWrapper"
                >
                  <img
                    alt="Your photo profile"
                    src="some_url"
                  />
                  <div
                    class="w-40 mt-2"
                  >
                    <button
                      class="Btn BtnDangerOutline"
                      data-testid="remove-photo-btn"
                      type="button"
                    >
                      Remove photo
                    </button>
                  </div>
                </div>
              </label>
            </div>
            <div
              class="w-4/5"
            >
              <h2>
                My Profile
              </h2>
              <ul>
                <li>
                  Name
                  : 
                  Adam
                </li>
                <li>
                  Lastname
                  : 
                  Smith
                </li>
                <li>
                  Email
                  : 
                  user@email.com
                </li>
                <li>
                  Username
                  : 
                  Adam123
                </li>
                <li>
                  Role
                  : 
                  Admin
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
